analytics.go

```diff
@@ -165,53 +171,120 @@ func (r *RedisAnalyticsHandler) Init() {
-		}
+	// start worker pool
+	atomic.SwapUint32(&r.shouldStop, 0)
+	for i := 0; i < ps; i++ {
+		r.poolWg.Add(1)
+		go r.recordWorker()
+	}
+}
 
-		if record.OrgID != "" {
-			record.Tags = append(record.Tags, "org-"+record.OrgID)
-		}
+func (r *RedisAnalyticsHandler) Stop() {
+	// flag to stop sending records into channel
+	atomic.SwapUint32(&r.shouldStop, 1)
 
-		record.Tags = append(record.Tags, "api-"+record.APIID)
+	// close channel to stop workers
+	close(r.recordsChan)
 
-		encoded, err := msgpack.Marshal(record)
+	// wait for all workers to be done
+	r.poolWg.Wait()
+}
 
-		if err != nil {
-			log.Error("Error encoding analytics data: ", err)
-		}
+// RecordHit will store an AnalyticsRecord in Redis
+func (r *RedisAnalyticsHandler) RecordHit(record *AnalyticsRecord) error {
+	// check if we should stop sending records 1st
+	if atomic.LoadUint32(&r.shouldStop) > 0 {
+		return nil
+	}
 
-		r.Store.AppendToSet(analyticsKeyName, string(encoded))
-	})
+	// just send record to channel consumed by pool of workers
+	// leave all data crunching and Redis I/O work for pool workers
+	r.recordsChan <- record
 
 	return nil
+}
+
+func (r *RedisAnalyticsHandler) recordWorker() {
+	defer r.poolWg.Done()
+
+	// this is buffer to send one pipelined command to redis
+	// use r.recordsBufferSize as cap to reduce slice re-allocations
+	recordsBuffer := make([]string, 0, r.recordsBufferSize)
+
+	// read records from channel and process
+	for {
+		readyToSend := false
+		select {
+
+		case record, ok := <-r.recordsChan:
+			// check if channel was closed and it is time to exit from worker
+			if !ok {
+				// send what is left in buffer
+				r.Store.AppendToSetPipelined(analyticsKeyName, recordsBuffer)
+				return
+			}
+
+			// we have new record - prepare it and add to buffer
+
+			// If we are obfuscating API Keys, store the hashed representation (config check handled in hashing function)
+			record.APIKey = storage.HashKey(record.APIKey)
+
+			if r.globalConf.SlaveOptions.UseRPC {
+				// Extend tag list to include this data so wecan segment by node if necessary
+				record.Tags = append(record.Tags, "tyk-hybrid-rpc")
+			}
+
+			if r.globalConf.DBAppConfOptions.NodeIsSegmented {
+				// Extend tag list to include this data so we can segment by node if necessary
+				record.Tags = append(record.Tags, r.globalConf.DBAppConfOptions.Tags...)
+			}
+
+			// Lets add some metadata
+			if record.APIKey != "" {
+				record.Tags = append(record.Tags, "key-"+record.APIKey)
+			}
+
+			if record.OrgID != "" {
+				record.Tags = append(record.Tags, "org-"+record.OrgID)
+			}
+
+			record.Tags = append(record.Tags, "api-"+record.APIID)
+
+			if encoded, err := msgpack.Marshal(record); err != nil {
+				log.WithError(err).Error("Error encoding analytics data")
+			} else {
+				recordsBuffer = append(recordsBuffer, string(encoded))
+			}
+
+			// identify that buffer is ready to be sent
+			readyToSend = uint64(len(recordsBuffer)) == r.recordsBufferSize
```

Comment:
Les say we have 50 workers, and 1000 buffer per worker and each receives traffic which has around 200ms around requests (but a bit lower), and do not hit timeout, mentioned above. Time to fill the buffer can be up to ((1000 * 50) * 0.2)/60 ~ 166 minutes. Correct me if I wrong. 

If this true, I suggest putting one more top timeout to this select, which will force writing analytics to redis every second.

Reply:
Thats true, current design assumes high traffic. but with low throughput like you said one request every 199ms - data will stale here for long time. We can add for that special case logic to send data to redis anyways if last time we sent data was more than 1 sec ago.